pkg/virt-config/config-map.go

```diff
@@ -427,35 +403,87 @@ func setConfig(config *Config, configMap *k8sv1.ConfigMap) error {
 // This will also allow us then to react to config changes and e.g. restart some controllers
-func (c *ClusterConfig) getConfig() (config *Config) {
+func (c *ClusterConfig) GetConfig() (config *v1.KubeVirtConfiguration) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
+	var configMap *k8sv1.ConfigMap
+	var kv *v1.KubeVirt
+	var resourceVersion string
+	useConfigMap := false
+
 	if obj, exists, err := c.configMapInformer.GetStore().GetByKey(c.namespace + "/" + ConfigMapName); err != nil {
 		log.DefaultLogger().Reason(err).Errorf("Error loading the cluster config from cache, falling back to last good resource version '%s'", c.lastValidConfig.ResourceVersion)
 		return c.lastValidConfig
 	} else if !exists {
-		return c.defaultConfig
+		kv = c.getConfigFromKubeVirtCR()
+		resourceVersion = kv.ResourceVersion
 	} else {
-		configMap := obj.(*k8sv1.ConfigMap)
-		if c.lastValidConfig.ResourceVersion == configMap.ResourceVersion ||
-			c.lastInvalidConfigResourceVersion == configMap.ResourceVersion {
-			return c.lastValidConfig
-		}
-		config := defaultClusterConfig()
-		if err := setConfig(config, configMap); err != nil {
-			c.lastInvalidConfigResourceVersion = configMap.ResourceVersion
-			log.DefaultLogger().Reason(err).Errorf("Invalid cluster config with resource version '%s', falling back to last good resource version '%s'", configMap.ResourceVersion, c.lastValidConfig.ResourceVersion)
-			return c.lastValidConfig
-		}
-		log.DefaultLogger().Infof("Updating cluster config to resource version '%s'", configMap.ResourceVersion)
-		c.lastValidConfig = config
+		useConfigMap = true
+		configMap = obj.(*k8sv1.ConfigMap)
+		resourceVersion = configMap.ResourceVersion
+	}
+
+	// if there is a configuration config map present we should use its configuration
+	// and ignore configuration in kubevirt
+	if c.lastValidConfig.ResourceVersion == resourceVersion ||
+		c.lastInvalidConfigResourceVersion == resourceVersion {
+		return c.lastValidConfig
+	}
+
+	config = defaultClusterConfig()
+	var err error
+	if useConfigMap {
+		err = setConfigFromConfigMap(config, configMap)
+	} else {
+		err = setConfigFromKubeVirt(config, kv)
+	}
+
+	if err != nil {
+		c.lastInvalidConfigResourceVersion = config.ResourceVersion
+		log.DefaultLogger().Reason(err).Errorf("Invalid cluster config with resource version '%s', falling back to last good resource version '%s'", config.ResourceVersion, c.lastValidConfig.ResourceVersion)
 		return c.lastValidConfig
 	}
+
+	log.DefaultLogger().Infof("Updating cluster config to resource version '%s'", config.ResourceVersion)
+	c.lastValidConfig = config
+	return c.lastValidConfig
+}
+
+func (c *ClusterConfig) getConfigFromKubeVirtCR() *v1.KubeVirt {
+	// TODO: I'm unsure how we know which kubevirt CR is the one in use
+	// once it is created it will have the status "DEPLOYED" but what about
+	// when it is still initalizing?
+	if obj, exists, err := c.kubeVirtInformer.GetStore().GetByKey(c.namespace + "/" + "kubevirt"); err != nil {
```

Comment:
Is there a way to know which KubeVirt CR creation triggered the creation of the `virt-api` and `virt-controller`? If so I could then pass that name to the `NewClusterConfig` along with the namespace. But i've been unable to figure out how?  or maybe there is a better idea for this?

Reply:
If there are multiple CRs, the kubevirt-operator should stop doing anything. Therefore you should at this stage only see one CR with a status which is indicating that it installed kubevirt. That is the right one. So the status on a second CR should pretty much not even show that it is initializing.